localstack/services/stepfunctions/asl/component/state/state_execution/execute_state.py

```diff
@@ -133,9 +157,43 @@ def _terminate_with_event(failure_event: FailureEvent, env: Environment) -> None
             ExecutionFailedEventDetails(**(list(failure_event.event_details.values())[0]))
         )
 
+    def _evaluate_with_timeout(self, env: Environment) -> None:
+        self.timeout.eval(env=env)
+        timeout_seconds: int = env.stack.pop()
+
+        frame: Environment = env.open_frame()
+        frame.inp = copy.deepcopy(env.inp)
+        frame.stack = copy.deepcopy(env.stack)
+        execution_outputs: list[Any] = list()
+        execution_exceptions: list[Optional[Exception]] = [None]
+        terminated_event = threading.Event()
+
+        def _exec_and_notify():
+            try:
+                self._eval_execution(frame)
+                execution_outputs.extend(frame.stack)
+            except Exception as ex:
+                execution_exceptions.append(ex)
+            terminated_event.set()
+
+        thread = Thread(target=_exec_and_notify)
+        thread.start()
+        finished_on_time: bool = terminated_event.wait(timeout_seconds)
```

Comment:
Need to be a bit more careful here regarding potential leaks.

It's not the nicest patterns but there's the `FuncThread` and/or `TMP_THREADS` concepts to keep track of started threads so that we can kill them on shutdown.

Let's keep it as is for now though and revisit the concurrency constructs in a separate project at a later stage :+1: